### PR TITLE
GPU: More robust error reporting

### LINKED ISSUE #10779: GPU: Functions that may return errors should call SDL_SetError()
Currently most (all?) of the error reporting is done through the SDL log, and APIs that may fail (e.g. `SDL_CreateGPUDevice`) do not call `SDL_SetError()`. This makes it impossible to, for example, display a message box with the failure reason.

### PR DESCRIPTION
**Pull request by thatcosmonaut**
Resolves #10779 

I rewrote a lot of error checking so that functions that can fail will bubble up the error to the caller. There is no longer a distinction between logging or setting an error because failures will always set an error. In debug mode the error function will also log.

Commands do not return errors. Since their execution is deferred, the expectation is that the graphics API's validator layer will catch those.

Most graphics errors are not sanely recoverable, but the client can now at least check for errors and report them before crashing.

Needs a Metal implementation before merge.

### CODE REVIEW

File: include/SDL3/SDL_gpu.h [hunk 1]
```diff
@@ -2301,7 +2302,7 @@ extern SDL_DECLSPEC void SDLCALL SDL_ReleaseGPUGraphicsPipeline(
  * acquired on.
  *
  * \param device a GPU context.
- * \returns a command buffer.
+ * \returns a command buffer, or NULL on failure.
```
Comment (slouken) [inline include/SDL3/SDL_gpu.h:2305]:
```suggestion
 * \returns a command buffer, or NULL on failure; call SDL_GetError() for more information.
```

File: include/SDL3/SDL_gpu.h [hunk 2]
```diff
@@ -2967,7 +2968,7 @@ extern SDL_DECLSPEC void SDLCALL SDL_EndGPUComputePass(
  * \param device a GPU context.
  * \param transfer_buffer a transfer buffer.
  * \param cycle if true, cycles the transfer buffer if it is already bound.
- * \returns the address of the mapped transfer buffer memory.
+ * \returns the address of the mapped transfer buffer memory, or NULL on failure.
```
Comment (slouken) [inline include/SDL3/SDL_gpu.h:2971]:
```suggestion
 * \returns the address of the mapped transfer buffer memory, or NULL on failure; call SDL_GetError() for more information.
```
...etc.

File: src/gpu/SDL_gpu.c [hunk 3]
```diff
@@ -2625,34 +2626,33 @@ SDL_GPUTexture *SDL_AcquireGPUSwapchainTexture(
...
     if (COMMAND_BUFFER_DEVICE->debug_mode) {
-        CHECK_COMMAND_BUFFER
+        CHECK_COMMAND_BUFFER_RETURN_FALSE
         if (
             commandBufferHeader->render_pass.in_progress ||
             commandBufferHeader->compute_pass.in_progress ||
             commandBufferHeader->copy_pass.in_progress) {
             SDL_assert_release(!"Cannot submit command buffer while a pass is in progress!");
-            return;
+            return false;
```
Comment (slouken) [inline src/gpu/SDL_gpu.c:2649]: This should set an error (and maybe log in debug mode instead of asserting?)

Comment (Akaricchi) [reply src/gpu/SDL_gpu.c:2649]: An assert is infinitely more useful here for debugging.

Comment (thatcosmonaut) [reply src/gpu/SDL_gpu.c:2649]: Our intention was to assert in gpu.c in debug mode because the errors we catch there are client misuses.

**Review by TheSpydog — APPROVED** [review]

**Comment by thatcosmonaut** [comment]
Currently making sure that we're logging errors in debug mode and then this is good to go.

**Comment by flibitijibibo** [comment]
Looks like Haiku CI went down... this should be clear to squash once it's back up and green.

**Pull request merged**